### PR TITLE
Menu API enhancements

### PR DESCRIPTION
**Pull request by nlemoine**

### CODE REVIEW

File: lib/Factory/MenuFactory.php
Comment (nlemoine) [inline lib/Factory/MenuFactory.php:110]: Don't throw exception for consistency with other "from" mtehods.

File: lib/Menu.php
Comment (nlemoine) [inline lib/Menu.php:236]: Caching menu plugins are [misusing this filter](https://github.com/inpsyde/menu-cache/blob/master/index.php#L43-L44) to cache menus and excepts a string here, provide a serialized menu object.

File: lib/Menu.php
Comment (nlemoine) [inline lib/Menu.php:254]: Make constructor protected to force factory usage.

File: lib/MenuItem.php
Comment (nlemoine) [inline lib/MenuItem.php:308]: Providing a Timber object in a core filter should not happen.

File: lib/MenuItem.php
Comment (gchtr) [inline lib/MenuItem.php:533]: Nice 😎

File: lib/Timber.php
Comment (gchtr) [inline lib/Timber.php:934]: I commented it before, but I would leave this as a separate function. See https://github.com/timber/timber/issues/2292 and https://github.com/timber/timber/issues/2089#issuecomment-628977116.

File: lib/Timber.php
Comment (gchtr) [inline lib/Timber.php:906]: I love this! 💚

**Review by gchtr — CHANGES_REQUESTED** [review]
I think you’re way deeper into the menu logic than I will ever be. Thanks a lot for all your work! There’s lots of optimizations in there. I also like the part about the caching compatibility with other plugins.

I saw a couple of things that I think we discussed before. I added comments right in the code.

File: lib/Factory/MenuFactory.php
Comment (nlemoine) [reply lib/Factory/MenuFactory.php:16]: The idea behind this was to mimic the `wp_nav_menu` behaviour which tries to find a menu by all means available.

Honestly, I'll personnaly never use that method so I'm fine with either option, just thought it would bring more parity with `wp_nav_menu`.

File: lib/Factory/MenuItemFactory.php
Comment (nlemoine) [reply lib/Factory/MenuItemFactory.php:24]: In my daily code practice, I find returning nullable types very handy. It allows null coalescing operator usage.

Let me know what's your option on this.

File: lib/Menu.php
Comment (nlemoine) [reply lib/Menu.php:153]: It follows the same idea mentionned in https://github.com/timber/timber/pull/2517#discussion_r762427431. `Menu::build` is basically `wp_nav_menu` adapted to Timber.

I read those threads about the PagesMenu API. Looks fine but I stayed in the current implementation scope. Those menus already bugged me enough :) I'll pass on this one.

File: lib/MenuItem.php
Comment (nlemoine) [reply lib/MenuItem.php:533]: I have mixed feelings about this and some parts of https://github.com/timber/timber/issues/1820#issue-370922635. 

Timber is actually trying to get rid of those filters everywhere to handle HTML. And menu are a noticeably awful example of this practice. I think we shouldn't put too much effort in being 100% compliant with all those filters (`wp_nav_menu_items`, `wp_nav_menu_{$menu->slug}_items`, etc.). It might not be possible or it might come at too much code costs. 

If some user wants full compatibility with all core menu filters, skipping Timber implemented menus and using `wp_nav_menu` is quite easy. I actually had an idea for this that I didn't commit in that PR that could be a fair compromise between all Timber menus VS `wp_nav_menu` only.

A `__toString` method on the `Menu` class that will dump the HTML menu using walker and all those core stuff:

```php
public function __toString() {
	static $menu_id_slugs = array();

	$args = $this->args;

	$items = '';
	$nav_menu = '';
	$show_container = false;

	if ( $args->container ) {
		/**
			* Filters the list of HTML tags that are valid for use as menu containers.
			*
			* @since 3.0.0
			*
			* @param string[] $tags The acceptable HTML tags for use as menu containers.
			*                       Default is array containing 'div' and 'nav'.
			*/
		$allowed_tags = apply_filters( 'wp_nav_menu_container_allowedtags', array( 'div', 'nav' ) );

		if ( is_string( $args->container ) && in_array( $args->container, $allowed_tags, true ) ) {
			$show_container = true;
			$class          = $args->container_class ? ' class="' . esc_attr( $args->container_class ) . '"' : ' class="menu-' . $this->slug . '-container"';
			$id             = $args->container_id ? ' id="' . esc_attr( $args->container_id ) . '"' : '';
			$aria_label     = ( 'nav' === $args->container && $args->container_aria_label ) ? ' aria-label="' . esc_attr( $args->container_aria_label ) . '"' : '';
			$nav_menu      .= '<' . $args->container . $id . $class . $aria_label . '>';
		}
	}

	$items .= walk_nav_menu_tree( $this->sorted_menu_items, $args->depth, $args );

	// Attributes.
	if ( ! empty( $args->menu_id ) ) {
		$wrap_id = $args->menu_id;
	} else {
		$wrap_id = 'menu-' . $this->slug;

		while ( in_array( $wrap_id, $menu_id_slugs, true ) ) {
			if ( preg_match( '#-(\d+)$#', $wrap_id, $matches ) ) {
				$wrap_id = preg_replace( '#-(\d+)$#', '-' . ++$matches[1], $wrap_id );
			} else {
				$wrap_id = $wrap_id . '-1';
			}
		}
	}
	$menu_id_slugs[] = $wrap_id;

	$wrap_class = $args->menu_class ? $args->menu_class : '';

	$nav_menu .= sprintf( $args->items_wrap, esc_attr( $wrap_id ), esc_attr( $wrap_class ), $items );
	if ( $show_container ) {
		$nav_menu .= '</' . $args->container . '>';
	}

	return $nav_menu;
}
```

So you could simply do `{{ header_menu }}` in your Twig template while still keeping the ability to use `Menu` objects/factories. 

Let me know if think this could be handy.

File: lib/Factory/MenuItemFactory.php
Comment (gchtr) [reply lib/Factory/MenuItemFactory.php]: Let’s use `null`. Being able to use the null coalescing operator is a big plus. In the meantime, I also saw it as a standard in a couple of other projects. WordPress’s `get_post()` function also returns `null` if no post can be found.

The menu functions are just not the only place we need to change this, though. I’ll create a separate issue for this.

File: lib/Factory/MenuFactory.php
Comment (gchtr) [reply lib/Factory/MenuFactory.php:16]: Yeah, that’s the way to usually do it. But because this has led to unexpected behaviors and problems, I’d like to follow what we discussed https://github.com/timber/timber/issues/2089#issuecomment-586739839 and implement the changes listed in #2292.

File: lib/MenuItem.php
Comment (gchtr) [reply lib/MenuItem.php:533]: That sounds like a great idea! I just wouldn’t use it with the `__toString()` method, but maybe add a `wp_nav_menu()` method?

```twig
{{ header_menu.wp_nav_menu }}
```

**Review by gchtr — APPROVED** [review]
I don’t want to keep this back much longer. There are lots of good things in here. I take three todos out of this:

- I’d like to just leave in the pages menu logic and refactor it with a separate issue (#2292).
- Then there’s the idea about the feature parity with `wp_nav_menu()`, which we found is quite hard to accomplish. A separate method that builds a menu using the nav menu walker is quite a good idea for people who need to use it for compatibility reasons (see https://github.com/timber/timber/pull/2517#discussion_r762430001).
- And finally, to return `null` instead of `false` from the factory methods so that they work with the null coalescing operator.

Once all the tests go green we can merge this in. Thank you so much for your work @nlemoine. Tell me if you still find the time to get this pull request ready more merging or if you’d like to pass it on.

**Comment by jarednova** [comment]
Just want to flag the two failing tests we have remaining ...

```
There were 2 failures:

1) TestMenuItemFactory::testMenuGarbageInGarbageOut
Failed asserting that false is null.

Error: /home/runner/work/timber/timber/tests/test-menu-item-factory.php:53
phpvfscomposer:///home/runner/work/timber/timber/vendor/phpunit/phpunit/phpunit:60

2) TestTimberTerm::testTermWithNativeMetaNotExisting
Failed asserting that 'qux' is null.

Error: /home/runner/work/timber/timber/tests/test-timber-term.php:4[49](https://github.com/timber/timber/runs/5343496381?check_suite_focus=true#step:15:49)
phpvfscomposer:///home/runner/work/timber/timber/vendor/phpunit/phpunit/phpunit:[60](https://github.com/timber/timber/runs/5343496381?check_suite_focus=true#step:15:60)```

File: lib/MenuItem.php
Comment (nlemoine) [reply lib/MenuItem.php:533]: I stand by the `__toString` method which makes more sense for the following reasons:
- The HTML menu is literally the string representation of the menu object
- The name `wp_nav_menu` is misleading, users could think it could take the same arguments than the native `wp_nav_menu` WordPress method, which is not

File: lib/Timber.php
Comment (nlemoine) [reply lib/Timber.php:937]: I saw that comment a couple of times :) I have no idea what that `PagesMenu` is about and [will leave that to someone else](https://github.com/timber/timber/pull/2517#discussion_r762428881) if you're ok with that.

**Comment by nlemoine** [comment]
@jarednova If fixed those, ACF finally rolled back the [`get_field` change](https://www.advancedcustomfields.com/changelog/).

@gchtr I added everything but the part concerning the PagesMenu API. Don't mind the commit messages, I'll squash everything with a cleaner one once validated :)

File: lib/MenuItem.php
Comment (nlemoine) [inline lib/MenuItem.php:115]: @gchtr @jarednova Is it really needed? This prevents nav walker to print menu titles.

File: lib/Factory/MenuItemFactory.php
Comment (gchtr) [reply lib/Factory/MenuItemFactory.php:24]: See #2561.

File: lib/Menu.php
Comment (gchtr) [reply lib/Menu.php:153]: > I read those threads about the PagesMenu API. Looks fine but I stayed in the current implementation scope. Those menus already bugged me enough :) I'll pass on this one.

All good. We can handle this in another pull request. I’ll try to see what I can do.

File: lib/MenuItem.php
Comment (gchtr) [reply lib/MenuItem.php:534]: @nlemoine I see. Yeah, that makes sense. Let’s go with `__toString()` then 🙂.

File: lib/MenuItem.php
Comment (gchtr) [reply lib/MenuItem.php:115]: Found the issue where the `$__title` property was introduced: https://github.com/timber/timber/issues/334.

The problem was when an ACF field named `title` existed. These kind of conflicts lead to the introduction of using `meta()` as the official way to get meta data from an object instead of directly using the field’s name. So if you have an ACF field for a menu item called title and you want to access it, you should use `{{ item.meta('title') }}`  instead of `{{ item.title }}`, because Timber wouldn’t know what to use.

We can definitely get rid of the `$__title` property and use `$title` directly.

Thanks for flagging this, @nlemoine!

File: lib/Factory/MenuFactory.php
Comment (gchtr) [inline lib/Factory/MenuFactory.php:52]: While working on the Pages Menu functionality (#2562) I stumbled over this again. And I propose to remove this if-statement.

If we leave this out and just return `null` when no menu couldn’t be determined up until here, then we wouldn’t run into unexpected behaviors where the wrong menu is returned (see #1440).

WordPress uses the following functionality in `wp_nav_menu()`:

```php
// Get the first menu that has items if we still can't find a menu.
if ( ! $menu && ! $args->theme_location ) {
    $menus = wp_get_nav_menus();
    foreach ( $menus as $menu_maybe ) {
        $menu_items = wp_get_nav_menu_items( $menu_maybe->term_id, array( 'update_post_term_cache' => false ) );
        if ( $menu_items ) {
            $menu = $menu_maybe;
            break;
        }
    }
}
```

We replicate this functionality in `MenuFactory::from_nav_menu_terms()`. This is fine if no parameter is passed to `Timber::get_menu()` by choice. However, if a parameter is passed, I think we should return `null` if no menu is found. This would be what I expect the function to do.

Since we have a lot of custom menu logic, I think we don’t have to follow what WordPress does here.

File: lib/Menu.php
Comment (gchtr) [inline lib/Menu.php:183]:
```suggestion
		/**
		 * Since Timber doesn’t use HTML, serialize the menu object to provide a cacheable string.
		 *
		 * Certain caching plugins will use this filter to cache a menu and return it early in the
		 * `pre_wp_nav_menu` filter.
		 *
		 * We can’t use the result of this filter, because it would return a string. That’s why we
		 * don’t assign the result of the filter to a variable.
		 *
		 * @see wp_nav_menu()
		 */
		apply_filters( 'wp_nav_menu', serialize( $nav_menu ), $args );
```

File: lib/Factory/MenuFactory.php
Comment (nlemoine) [reply lib/Factory/MenuFactory.php:52]: Yes, I fully agree! I will update this.

File: tests/test-menu-item-factory.php
Comment (nlemoine) [inline tests/test-menu-item-factory.php:173]: @gchtr This comment may aloso apply to other classmaps tests

File: lib/MenuItem.php
Comment (nlemoine) [inline lib/MenuItem.php:17]: Update with proposed name (@gchtr in #2560)

File: lib/Factory/MenuItemFactory.php
Comment (nlemoine) [inline lib/Factory/MenuItemFactory.php:66]: While working with 2.x on a project, I found that it could be quite useful to have nav location based classmaps for menu items.

File: lib/Factory/MenuFactory.php
Comment (gchtr) [reply lib/Factory/MenuFactory.php]: I think now the logic is still the same, just switched around.

I propose that we don’t call `$this->from_nav_menu_terms($args)` in `MenuFactory::from()` if no menu can be found. With the current logic, if no menu could be found with the provided args, then Timber will *still* try to return the first menu it can find.

So instead of …

```php
if($menu) {
    return $menu;
}

return $this->from_nav_menu_terms($args);
```

at the end of the method, I would just do …

```php
return $menu;
```

This would fix the bug in #1440 and would be more in line with what I would expect the function to do.

File: lib/MenuItem.php
Comment (gchtr) [reply lib/MenuItem.php:17]: Thanks!

**Review by gchtr — CHANGES_REQUESTED** [review]
Nice, thanks for the additional updates @nlemoine. I like the new `timber/menuitem/classmap` filter 💪. And good catch with the `assertInstanceOf()` checks. I added two more issues to finish work on these in #2571 and #2572.

For me, the only remaining discussion is in https://github.com/timber/timber/pull/2517#discussion_r848621890. Apart from that it would be nice if we could merge this in, soon.

If you want to add additional features, I think it would be nice to add it in a separate pull request. I think it’s better to have many smaller pull requests instead of having one big pull request that keeps growing and growing.

**Comment by gchtr** [comment]
Oh, and https://github.com/timber/timber/pull/2517#discussion_r820128943 is also unfinished 🙂. Do we want to move that to a separate issue as well, or do you want to solve it here, @nlemoine?

File: lib/MenuItem.php
Comment (nlemoine) [reply lib/MenuItem.php:115]: Removing it is not so easy, because we need Twig to call the `title()` method in order to apply the `nav_menu_item_title`. If `$item->title` becomes a public property, `{{ item.title }}` won't call `title()` and no filter will be applied.

I added a `$title` and marked it as protected to solve this.

File: lib/Factory/MenuFactory.php
Comment (nlemoine) [reply lib/Factory/MenuFactory.php]: Sorry, this was not unfinished, this is now fixed!

**Comment by nlemoine** [comment]
@gchtr I think we're finally good on menus 😅 

Please check my comment here: https://github.com/timber/timber/pull/2517#discussion_r861541645 for the `title` explanation.

File: lib/MenuItem.php
Comment (gchtr) [reply lib/MenuItem.php]: @nlemoine Thanks for the update! Would it make sense to completely get rid of the `$title` property? Could this be a use case for #2560?

If we use `{{ item.title }}`, then the `title()` method would be called, were we could use `$this->wp_object->title` instead of `$this->title`:

```php
$title = apply_filters( 'nav_menu_item_title', $this->wp_object->title, $this->wp_object, $this->menu->args ? $this->menu->args : new \stdClass, $this->level );
```

Or might I be missing something?

Comment (gchtr) [reply lib/MenuItem.php]: Good point! I see you’ve thought this through. Both solutions are fine for me then, too.

**Review by gchtr — APPROVED** [review]

File: lib/MenuItem.php
Comment (nlemoine) [reply lib/MenuItem.php]: So change is required I guess :) Let me know if you still see something, the PR is ready to merge from my side.

**Comment by nlemoine** [comment]
Do I still need to wait for @jarednova approval?

**Comment by gchtr** [comment]
@nlemoine No, I think Jared would have commented already earlier in the process if he saw something. You can merge it in 🎉. Thanks for your work on this!

**Pull request merged**
